File: sdk/typescript/src/context.ts
/**
 * Context propagation using AsyncLocalStorage.
 */

import { AsyncLocalStorage } from "async_hooks";
import type { AgentTrace, Trace, Span, Generation } from "./client";

// AsyncLocalStorage for context propagation
const clientStorage = new AsyncLocalStorage<AgentTrace | null>();
const traceStorage = new AsyncLocalStorage<Trace | null>();
const observationStorage = new AsyncLocalStorage<Span | Generation | null>();

// Fallback for non-async contexts
let globalClient: AgentTrace | null = null;
let globalTrace: Trace | null = null;
let globalObservation: Span | Generation | null = null;

/**
 * Set the global AgentTrace client.
 */
export function setClient(client: AgentTrace | null): void {
  globalClient = client;
}

/**
 * Get the global AgentTrace client.
 */
export function getClient(): AgentTrace | null {
  return clientStorage.getStore() ?? globalClient;
}

/**
 * Set the current trace in context.
 */
export function setCurrentTrace(trace: Trace | null): void {
  globalTrace = trace;
}

/**
 * Get the current trace from context.
 */
export function getCurrentTrace(): Trace | null {
  return traceStorage.getStore() ?? globalTrace;
}

/**
 * Set the current observation in context.
 */
export function setCurrentObservation(observation: Span | Generation | null): void {
  globalObservation = observation;
}

/**
 * Get the current observation from context.
 */
export function getCurrentObservation(): Span | Generation | null {
  return observationStorage.getStore() ?? globalObservation;
}

/**
 * Run a function within a trace context.
 *
 * @example
 * ```typescript
 * const result = await runWithTrace(trace, async () => {
 *   // Code here has access to the trace via getCurrentTrace()
 *   return await doSomething();
 * });
 * ```
 */
export function runWithTrace<T>(trace: Trace, fn: () => T): T {
  return traceStorage.run(trace, fn);
}

/**
 * Run a function within an observation context.
 *
 * @example
 * ```typescript
 * const result = await runWithObservation(span, async () => {
 *   // Code here has access to the observation via getCurrentObservation()
 *   return await doSomething();
 * });
 * ```
 */
export function runWithObservation<T>(observation: Span | Generation, fn: () => T): T {
  return observationStorage.run(observation, fn);
}

/**
 * Run a function within a client context.
 */
export function runWithClient<T>(client: AgentTrace, fn: () => T): T {
  return clientStorage.run(client, fn);
}

/**
 * Context wrapper class for managing trace context.
 */
export class TraceContext {
  private trace: Trace;
  private previousTrace: Trace | null = null;

  constructor(trace: Trace) {
    this.trace = trace;
  }

  /**
   * Enter the context.
   */
  enter(): Trace {
    this.previousTrace = globalTrace;
    globalTrace = this.trace;
    return this.trace;
  }

  /**
   * Exit the context.
   */
  exit(): void {
    globalTrace = this.previousTrace;
    this.previousTrace = null;
  }

  /**
   * Run a function within this trace context.
   */
  run<T>(fn: () => T): T {
    this.enter();
    try {
      return fn();
    } finally {
      this.exit();
    }
  }

  /**
   * Run an async function within this trace context.
   */
  async runAsync<T>(fn: () => Promise<T>): Promise<T> {
    return runWithTrace(this.trace, fn);
  }
}

/**
 * Context wrapper class for managing observation context.
 */
export class ObservationContext {
  private observation: Span | Generation;
  private previousObservation: Span | Generation | null = null;

  constructor(observation: Span | Generation) {
    this.observation = observation;
  }

  /**
   * Enter the context.
   */
  enter(): Span | Generation {
    this.previousObservation = globalObservation;
    globalObservation = this.observation;
    return this.observation;
  }

  /**
   * Exit the context.
   */
  exit(): void {
    globalObservation = this.previousObservation;
    this.previousObservation = null;
  }

  /**
   * Run a function within this observation context.
   */
  run<T>(fn: () => T): T {
    this.enter();
    try {
      return fn();
    } finally {
      this.exit();
    }
  }

  /**
   * Run an async function within this observation context.
   */
  async runAsync<T>(fn: () => Promise<T>): Promise<T> {
    return runWithObservation(this.observation, fn);
  }
}

File: sdk/python/agenttrace/context.py
"""
Context propagation for tracing.
"""

from __future__ import annotations

import contextvars
from typing import TYPE_CHECKING, Optional

if TYPE_CHECKING:
    from agenttrace.client import AgentTrace, Trace, Span, Generation

# Context variables for thread-safe and async-safe context propagation
_client_var: contextvars.ContextVar[Optional["AgentTrace"]] = contextvars.ContextVar(
    "agenttrace_client", default=None
)
_trace_var: contextvars.ContextVar[Optional["Trace"]] = contextvars.ContextVar(
    "agenttrace_trace", default=None
)
_observation_var: contextvars.ContextVar[Optional["Span | Generation"]] = contextvars.ContextVar(
    "agenttrace_observation", default=None
)


def set_client(client: Optional["AgentTrace"]) -> None:
    """Set the global AgentTrace client."""
    _client_var.set(client)


def get_client() -> Optional["AgentTrace"]:
    """Get the global AgentTrace client."""
    return _client_var.get()


def set_current_trace(trace: Optional["Trace"]) -> None:
    """Set the current trace in context."""
    _trace_var.set(trace)


def get_current_trace() -> Optional["Trace"]:
    """Get the current trace from context."""
    return _trace_var.get()


def set_current_observation(observation: Optional["Span | Generation"]) -> None:
    """Set the current observation in context."""
    _observation_var.set(observation)


def get_current_observation() -> Optional["Span | Generation"]:
    """Get the current observation from context."""
    return _observation_var.get()


class TraceContext:
    """
    Context manager for scoping a trace.

    Example:
        with TraceContext(trace):
            # Code here runs with trace as current context
            pass
    """

    def __init__(self, trace: "Trace"):
        self.trace = trace
        self._token: Optional[contextvars.Token] = None

    def __enter__(self) -> "Trace":
        self._token = _trace_var.set(self.trace)
        return self.trace

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        if self._token is not None:
            _trace_var.reset(self._token)


class ObservationContext:
    """
    Context manager for scoping an observation.

    Example:
        with ObservationContext(span):
            # Code here runs with span as current observation
            pass
    """

    def __init__(self, observation: "Span | Generation"):
        self.observation = observation
        self._token: Optional[contextvars.Token] = None

    def __enter__(self) -> "Span | Generation":
        self._token = _observation_var.set(self.observation)
        return self.observation

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        if self._token is not None:
            _observation_var.reset(self._token)


def run_in_trace_context(trace: "Trace", func, *args, **kwargs):
    """
    Run a function within a trace context.

    This is useful for running callbacks or parallel tasks while
    maintaining the trace context.

    Example:
        def my_callback():
            # This runs in the trace context
            current = get_current_trace()

        run_in_trace_context(trace, my_callback)
    """
    token = _trace_var.set(trace)
    try:
        return func(*args, **kwargs)
    finally:
        _trace_var.reset(token)


async def run_in_trace_context_async(trace: "Trace", coro):
    """
    Run an async coroutine within a trace context.

    Example:
        async def my_async_task():
            current = get_current_trace()

        await run_in_trace_context_async(trace, my_async_task())
    """
    token = _trace_var.set(trace)
    try:
        return await coro
    finally:
        _trace_var.reset(token)
